### ISSUE TITLE
List.fill performance problem

### ISSUE THREAD

**Issue by shawjef3** [scala/collection-strawman#233]
#232 shows that List.fill has a severe performance problem for large values of `n`. I suspect other builder functions have the same problem.

```
[info] Benchmark                 (size)  Mode  Cnt           Score            Error  Units
[info] ListBenchmark.fill       7312102  avgt   10  1016141204.700 ± 3015303989.604  ns/op
[info] ScalaListBenchmark.fill  7312102  avgt   10    72150762.767 ±    8988583.569  ns/op
```

**Comment by EPronovost** [comment]
From what I looked at, it appears that the generic `List.fill` falls back to `SeqFactor.fill`, which calls `fromIterable`.  This, in turn, uses a `ListBuilder` and `+= elem` _n_ times.  For Lists, this is a lot of extra baggage.  I wrote an optimized version of `fill` for `List` and was able to increase performance.  Here's the commit: https://github.com/EPronovost/collection-strawman/commit/aa5829efc6628041b5c7df6abf96e502e730ad2b

```
[info] Benchmark                    (size)  Mode  Cnt          Score          Error  Units
[info] FillBenchmark.fillExisting  7312102  avgt   12  131716487.886 ± 36529591.411  ns/op
[info] FillBenchmark.fillNew       7312102  avgt   12   77262934.910 ± 28352202.580  ns/op
[info] FillBenchmark.fillScala     7312102  avgt   12   92711301.221 ± 27794612.476  ns/op
```

**Comment by shawjef3** [comment]
That's a great result. We'll want to use ListBuffer though, so that `fill` appends rather than prepends.

**Comment by shawjef3** [comment]
I'm just learning this code, but it looks like the intended way to optimize this is to use SeqFactory.Delegate. However, an object can't be its own class constructor argument. IE we can't do `object List extends SeqFactory.Delegate[List](this)`. Maybe it's time to consider using a type class instead.

**Comment by shawjef3** [comment]
`fill` uses `ListBuffer.+=`, which should be quite fast. It's how the current collections work, also.

**Comment by shawjef3** [comment]
Attaching VisualVM cpu profiling results for `strawman.collection.immutable.List.fill(7312102)(())`.
![list fill](https://user-images.githubusercontent.com/1625822/30433280-0f266678-9932-11e7-83e3-22d705fbfc24.png)

**Comment by shawjef3** [comment]
Attaching VisualVM cpu profiling results for scala.List.fill(7312102)(()). I stopped the profiling early to avoid profiling toString(), so the invocation counts will be low.
![scala list fill](https://user-images.githubusercontent.com/1625822/30433790-65449e84-9933-11e7-96de-ed4bd2847220.png)

**Comment by shawjef3** [comment]
I found that part of the problem is the default isEmpty implementation for Iterable. Will post a patch later that gets strawman.List.fill within an order of magnitude of List.fill.

**Comment by shawjef3** [comment]
This problem isn't as severe as I thought. Both implementations can jump two orders of magnitude between iterations depending on what the GC is doing.

ListBenchmark.fill example with default GC:
```
[info] # Warmup Iteration   4: 68210054.500 ns/op
[info] # Warmup Iteration   5: 7010360800.000 ns/op
```

ScalaListBenchmark.fill example with default GC:
```
[info] # Warmup Iteration   4: 68144053.500 ns/op
[info] # Warmup Iteration   5: 6933620022.000 ns/op
```

If I use G1GC, the problem largely goes away. However, ScalaList still has a small advantage. I found that this is due to Iterable's isEmpty implementation creating an Iterator for each call, even when `knownSize > -1`.

**Comment by shawjef3** [comment]
I found that the huge variation between iterations is fixed by adding "-gc true" to the jmh arguments.

**Comment by shawjef3** [comment]
It turns out this GC turn is because of
```scala
  @Setup(Level.Invocation)
  def initInvocation(): Unit
```
Setting it to Level.Iteration really fixes this. Turns out that was the entire problem.

Look how iterations can vary by two orders of magnitude.

```
[info] # Warmup Iteration   1: 5772986070.000 ns/op
[info] # Warmup Iteration   2: 68278620.000 ns/op
[info] # Warmup Iteration   3: 6646371053.000 ns/op
[info] # Warmup Iteration   4: 65177938.000 ns/op
[info] # Warmup Iteration   5: 65828715.000 ns/op
[info] # Warmup Iteration   6: 6352249764.000 ns/op
[info] # Warmup Iteration   7: 73122346.000 ns/op
[info] # Warmup Iteration   8: 4751031310.000 ns/op
[info] # Warmup Iteration   9: 64778512.000 ns/op
[info] # Warmup Iteration  10: 65634132.000 ns/op
[info] # Warmup Iteration  11: 4516350996.000 ns/op
[info] # Warmup Iteration  12: 71127951.000 ns/op
```

Now look how much more consistent the results are after the change.

```
[info] # Warmup Iteration   1: 2397066201.000 ns/op
[info] # Warmup Iteration   2: 1934948080.000 ns/op
[info] # Warmup Iteration   3: 1413323666.500 ns/op
[info] # Warmup Iteration   4: 931746515.000 ns/op
[info] # Warmup Iteration   5: 203033997.400 ns/op
[info] # Warmup Iteration   6: 6204619049.000 ns/op
[info] # Warmup Iteration   7: 3238766224.500 ns/op
[info] # Warmup Iteration   8: 1852228277.000 ns/op
[info] # Warmup Iteration   9: 2442146471.000 ns/op
[info] # Warmup Iteration  10: 1976784038.000 ns/op
[info] # Warmup Iteration  11: 1734129563.667 ns/op
[info] # Warmup Iteration  12: 3211556343.000 ns/op
[info] Iteration   1: 1871131599.333 ns/op
[info] Iteration   2: 2500839376.000 ns/op
[info] Iteration   3: 2007281320.000 ns/op
[info] Iteration   4: 1766301722.667 ns/op
[info] Iteration   5: 3207502065.000 ns/op
[info] Iteration   6: 1889539856.333 ns/op
[info] Iteration   7: 2497703641.500 ns/op
[info] Iteration   8: 2024393950.000 ns/op
[info] Iteration   9: 1772549254.333 ns/op
[info] Iteration  10: 3240170542.500 ns/op
[info] Iteration  11: 1918782890.000 ns/op
[info] Iteration  12: 2566992625.000 ns/op
```

These were the List.fill benchmark ran without any options.

**Comment by odd** [comment]
Hi @shawjef3, the `@Level.Invocation` was added by me (in #202, now merged) to overcome another problem with using the same collection instances for multiple benchmark invocation (i.e. the invocations were not independent). With `Level.Invocation` each benchmark invocation gets its own fresh collection instance which avoids the collection sharing issue but instead does lead to more garbage being generated of course. 

You mentioned earlier that adding `-gc true` removed the inconsistency between iterations, is this still the case or is going back to `Level.Iteration` the only way of making the iteration times consistent again? If `-gc true` does not fully solve the issue we should probably go back to using `Level.Iteration` for most of the benchmarks and instead add specific benchmarks with `Level.Invocation` for the cases where `Level.Iteration` leads to skewed results caused by collection instance sharing.

**Comment by shawjef3** [comment]
@odd I looked at some more benchmarks and found some interesting things. Is GC overhead why you have `ListBenchmark#create` run 100 times in a loop? I can see that as a way to try to abstract over garbage collection.

**Comment by shawjef3** [comment]
After making ListBenchmark.fill and ScalaListBenchmark.fill loop 100 times per invocation, I found that List.fill is marginally slower than ScalaList.fill, but the huge performance deficit I thought was there was completely due to how the garbage collections happened.

This is not the big deal I thought it was.

```
[info] Benchmark                 (size)  Mode  Cnt          Score          Error  Units
[info] ListBenchmark.fill       7312102  avgt   12  214603357.384 ± 36347928.562  ns/op
[info] ScalaListBenchmark.fill  7312102  avgt   12  184305876.061 ± 22657902.609  ns/op
```

**Comment by shawjef3** [comment]
Benchmarks are so fickle. Here's what happens if I run the above benchmark with the serial garbage collector. Now `List.fill` is significantly slower.

```
[info] Benchmark                 (size)  Mode  Cnt          Score          Error  Units
[info] ListBenchmark.fill       7312102  avgt   12  130191466.283 ± 14611260.067  ns/op
[info] ScalaListBenchmark.fill  7312102  avgt   12   84734959.263 ± 16570560.271  ns/op
```

**Comment by shawjef3** [comment]
Well anyway, what I've learned is that preventing allocation of an Iterator during ListBuffer#isEmpty doesn't help. I was thinking that taking advantage of `knownSize` for `isEmpty` instead of allocating an iterator might provide some benefit, but it doesn't seem to.

```[info] Benchmark                       (size)  Mode  Cnt  Score   Error  Units
[info] ListBufferBenchmark.isEmptyNew       1  avgt   12  5.398 ▒ 0.030  ns/op
[info] ListBufferBenchmark.isEmptyOld       1  avgt   12  5.406 ▒ 0.036  ns/op```

This is what I added to Iterable.scala to benchmark the above.

```scala
  /** Is the collection empty? */
  def isEmpty: Boolean = {
    //Avoid creating an iterator if we already know the size.
    val computedKnownSize: Int = knownSize
    if (computedKnownSize > -1)
      computedKnownSize == 0
    else !toIterable.iterator().hasNext
  }

  /** Is the collection not empty? */
  def nonEmpty: Boolean = !isEmpty
````

**Comment by odd** [comment]
@shawjef3 The reason for doing the actual `create` operation in a loop was that when using `Level.Invocation` the initialization time of the collection fields would be large enough in relation to the actual benchmark invocation time to skew the results. But if we change back to using `Level.Trial` we could simplify the benchmarks and remove the looping. I've created a PR for that: #245.

**Comment by shawjef3** [comment]
I made isEmpty way too complicated. This is simpler, but still slower.

```scala
  def isEmpty: Boolean = {
    //Avoid creating an iterator if we already know the size.
    knownSize == 0 || !toIterable.iterator().hasNext
  }
```

**Comment by julienrf** [comment]
@shawjef3 IIUC that comment https://github.com/scala/collection-strawman/issues/233#issuecomment-329861385 the performance regression is not that huge?

I think we should update the `isEmpty` implementation in `IterableOnce` to what you’ve written above, even though it didn’t show a significant improvement in that benchmark.

**Comment by julienrf** [comment]
Hey!

I’ve run the benchmark given in #232 and we effectively are between 2x to 20x slower than the current collections. Here are the detailed result:

![fill](https://user-images.githubusercontent.com/332812/31725858-aad122a8-b425-11e7-9dea-b23f6c62f993.png)

~~~
[info] Benchmark                      (size)  Mode  Cnt         Score          Error  Units
[info] ListFillBenchmarks.fill             0  avgt    8        63.071 ±       16.316  ns/op
[info] ListFillBenchmarks.fill             1  avgt    8        62.975 ±        1.338  ns/op
[info] ListFillBenchmarks.fill             2  avgt    8        70.110 ±        0.903  ns/op
[info] ListFillBenchmarks.fill             3  avgt    8        76.061 ±        0.617  ns/op
[info] ListFillBenchmarks.fill             4  avgt    8        82.497 ±        0.849  ns/op
[info] ListFillBenchmarks.fill             7  avgt    8       102.004 ±        2.411  ns/op
[info] ListFillBenchmarks.fill             8  avgt    8       108.855 ±        6.923  ns/op
[info] ListFillBenchmarks.fill            15  avgt    8       153.340 ±        5.451  ns/op
[info] ListFillBenchmarks.fill            16  avgt    8       159.890 ±        3.992  ns/op
[info] ListFillBenchmarks.fill            17  avgt    8       166.250 ±        2.695  ns/op
[info] ListFillBenchmarks.fill            39  avgt    8       312.888 ±        6.620  ns/op
[info] ListFillBenchmarks.fill           282  avgt    8      1888.001 ±       98.034  ns/op
[info] ListFillBenchmarks.fill          4096  avgt    8     26378.845 ±      383.970  ns/op
[info] ListFillBenchmarks.fill        131070  avgt    8    894854.368 ±    57599.780  ns/op
[info] ListFillBenchmarks.fill       7312102  avgt    8  88026649.694 ± 21819653.519  ns/op
[info] ScalaListFillBenchmarks.fill        0  avgt    8         3.160 ±        0.464  ns/op
[info] ScalaListFillBenchmarks.fill        1  avgt    8         7.040 ±        0.492  ns/op
[info] ScalaListFillBenchmarks.fill        2  avgt    8        11.110 ±        0.449  ns/op
[info] ScalaListFillBenchmarks.fill        3  avgt    8        14.947 ±        0.498  ns/op
[info] ScalaListFillBenchmarks.fill        4  avgt    8        18.722 ±        0.503  ns/op
[info] ScalaListFillBenchmarks.fill        7  avgt    8        30.299 ±        0.747  ns/op
[info] ScalaListFillBenchmarks.fill        8  avgt    8        33.845 ±        0.291  ns/op
[info] ScalaListFillBenchmarks.fill       15  avgt    8        61.261 ±        0.336  ns/op
[info] ScalaListFillBenchmarks.fill       16  avgt    8        67.908 ±        5.099  ns/op
[info] ScalaListFillBenchmarks.fill       17  avgt    8        70.202 ±        1.357  ns/op
[info] ScalaListFillBenchmarks.fill       39  avgt    8       164.304 ±        6.854  ns/op
[info] ScalaListFillBenchmarks.fill      282  avgt    8      1168.763 ±       47.152  ns/op
[info] ScalaListFillBenchmarks.fill     4096  avgt    8     15938.166 ±      201.876  ns/op
[info] ScalaListFillBenchmarks.fill   131070  avgt    8    450498.820 ±    76816.504  ns/op
[info] ScalaListFillBenchmarks.fill  7312102  avgt    8  71174568.325 ± 21687327.564  ns/op
~~~

One main difference between the current `List.fill` method and the new `List.fill` method is that the former is builder-based whereas the latter is view-based. Being view based means that we allocate a `View`, we get its `Iterator`, and we loop through it to copy its elements to a `ListBuffer`, which we eventually convert to a `List` (this is a O(1) operation). The builder-based version runs a while loop (with no calls to `hasNext` or `next()` in each iteration) to put the elements in a builder on which it eventually calls `result()`.

If we switch to a builder-based version, we are “only” between 1.8x to 2x slower. Here are the detailed results:

![fill](https://user-images.githubusercontent.com/332812/31725320-54124ec0-b424-11e7-8b98-95271a3ffe8a.png)

~~~
[info] Benchmark                      (size)  Mode  Cnt          Score          Error  Units
[info] ListFillBenchmarks.fill             0  avgt    8          6.155 ±        0.042  ns/op
[info] ListFillBenchmarks.fill             1  avgt    8         12.885 ±        0.071  ns/op
[info] ListFillBenchmarks.fill             2  avgt    8         23.851 ±        9.650  ns/op
[info] ListFillBenchmarks.fill             3  avgt    8         28.434 ±        0.260  ns/op
[info] ListFillBenchmarks.fill             4  avgt    8         36.062 ±        0.264  ns/op
[info] ListFillBenchmarks.fill             7  avgt    8         58.790 ±        0.595  ns/op
[info] ListFillBenchmarks.fill             8  avgt    8         66.179 ±        0.565  ns/op
[info] ListFillBenchmarks.fill            15  avgt    8        120.001 ±        2.231  ns/op
[info] ListFillBenchmarks.fill            16  avgt    8        126.864 ±        1.074  ns/op
[info] ListFillBenchmarks.fill            17  avgt    8        134.766 ±        1.768  ns/op
[info] ListFillBenchmarks.fill            39  avgt    8        306.982 ±        5.840  ns/op
[info] ListFillBenchmarks.fill           282  avgt    8       2156.660 ±       45.336  ns/op
[info] ListFillBenchmarks.fill          4096  avgt    8      30987.071 ±      425.642  ns/op
[info] ListFillBenchmarks.fill        131070  avgt    8     788364.790 ±    20478.630  ns/op
[info] ListFillBenchmarks.fill       7312102  avgt    8  117394249.156 ± 24024250.783  ns/op
[info] ScalaListFillBenchmarks.fill        0  avgt    8          2.962 ±        0.019  ns/op
[info] ScalaListFillBenchmarks.fill        1  avgt    8          6.551 ±        0.035  ns/op
[info] ScalaListFillBenchmarks.fill        2  avgt    8         10.778 ±        0.155  ns/op
[info] ScalaListFillBenchmarks.fill        3  avgt    8         14.664 ±        0.154  ns/op
[info] ScalaListFillBenchmarks.fill        4  avgt    8         18.579 ±        0.348  ns/op
[info] ScalaListFillBenchmarks.fill        7  avgt    8         30.252 ±        0.211  ns/op
[info] ScalaListFillBenchmarks.fill        8  avgt    8         34.273 ±        0.955  ns/op
[info] ScalaListFillBenchmarks.fill       15  avgt    8        107.036 ±        6.590  ns/op
[info] ScalaListFillBenchmarks.fill       16  avgt    8         75.901 ±        0.301  ns/op
[info] ScalaListFillBenchmarks.fill       17  avgt    8         79.008 ±       11.629  ns/op
[info] ScalaListFillBenchmarks.fill       39  avgt    8        164.342 ±        4.242  ns/op
[info] ScalaListFillBenchmarks.fill      282  avgt    8       1126.003 ±       32.348  ns/op
[info] ScalaListFillBenchmarks.fill     4096  avgt    8      15757.792 ±      127.241  ns/op
[info] ScalaListFillBenchmarks.fill   131070  avgt    8     426613.585 ±     4512.173  ns/op
[info] ScalaListFillBenchmarks.fill  7312102  avgt    8   64553320.047 ± 15811815.206  ns/op
~~~

I guess one reason to still be behind the current collections is that our `List` builder is a wrapper around a `ListBuffer` whereas in the current collections, `ListBuffer` **is** a `List` builder (so it removes one level of indirection).

If we make `ListBuffer[A]` a `Builder[A, List[A]]`, then we become slightly closer to the current collection’s performance, but still between 1.15x to 1.6x slower:

![fill](https://user-images.githubusercontent.com/332812/31727858-053565ba-b42b-11e7-8990-5433622abcd1.png)

~~~
[info] Benchmark                      (size)  Mode  Cnt         Score          Error  Units
[info] ListFillBenchmarks.fill             0  avgt    8         2.958 ±        0.018  ns/op
[info] ListFillBenchmarks.fill             1  avgt    8         8.659 ±        0.054  ns/op
[info] ListFillBenchmarks.fill             2  avgt    8        15.652 ±        0.196  ns/op
[info] ListFillBenchmarks.fill             3  avgt    8        21.528 ±        0.153  ns/op
[info] ListFillBenchmarks.fill             4  avgt    8        28.154 ±        0.285  ns/op
[info] ListFillBenchmarks.fill             7  avgt    8        47.948 ±        0.405  ns/op
[info] ListFillBenchmarks.fill             8  avgt    8        56.171 ±        2.883  ns/op
[info] ListFillBenchmarks.fill            15  avgt    8       101.886 ±        1.919  ns/op
[info] ListFillBenchmarks.fill            16  avgt    8       108.414 ±        3.938  ns/op
[info] ListFillBenchmarks.fill            17  avgt    8       115.572 ±        5.349  ns/op
[info] ListFillBenchmarks.fill            39  avgt    8       269.576 ±       16.616  ns/op
[info] ListFillBenchmarks.fill           282  avgt    8      1882.431 ±       53.435  ns/op
[info] ListFillBenchmarks.fill          4096  avgt    8     27177.490 ±      368.861  ns/op
[info] ListFillBenchmarks.fill        131070  avgt    8    784631.266 ±    10743.474  ns/op
[info] ListFillBenchmarks.fill       7312102  avgt    8  66293246.112 ± 37729502.445  ns/op
[info] ScalaListFillBenchmarks.fill        0  avgt    8         3.042 ±        0.058  ns/op
[info] ScalaListFillBenchmarks.fill        1  avgt    8         6.628 ±        0.200  ns/op
[info] ScalaListFillBenchmarks.fill        2  avgt    8        10.865 ±        0.575  ns/op
[info] ScalaListFillBenchmarks.fill        3  avgt    8        14.787 ±        0.171  ns/op
[info] ScalaListFillBenchmarks.fill        4  avgt    8        18.594 ±        0.216  ns/op
[info] ScalaListFillBenchmarks.fill        7  avgt    8        30.438 ±        0.536  ns/op
[info] ScalaListFillBenchmarks.fill        8  avgt    8        35.216 ±        3.655  ns/op
[info] ScalaListFillBenchmarks.fill       15  avgt    8        62.409 ±        2.968  ns/op
[info] ScalaListFillBenchmarks.fill       16  avgt    8        67.442 ±        3.715  ns/op
[info] ScalaListFillBenchmarks.fill       17  avgt    8        75.503 ±        4.087  ns/op
[info] ScalaListFillBenchmarks.fill       39  avgt    8       163.350 ±        9.226  ns/op
[info] ScalaListFillBenchmarks.fill      282  avgt    8      1114.375 ±       25.540  ns/op
[info] ScalaListFillBenchmarks.fill     4096  avgt    8     16795.959 ±     2800.531  ns/op
[info] ScalaListFillBenchmarks.fill   131070  avgt    8    449145.273 ±    39378.066  ns/op
[info] ScalaListFillBenchmarks.fill  7312102  avgt    8  61872483.394 ± 16982269.950  ns/op
~~~

I will continue to investigate further…

**Comment by julienrf** [comment]
I consider that this issue has been fixed by #271.
